gcc/d/d-convert.cc

```diff
@@ -133,11 +143,13 @@ d_truthvalue_conversion (tree expr)
     case ADDR_EXPR:
       /* If we are taking the address of a decl that can never be null,
 	 then the return result is always true.  */
-      if (DECL_P (TREE_OPERAND (expr, 0))
-	  && (TREE_CODE (TREE_OPERAND (expr, 0)) == PARM_DECL
-	      || TREE_CODE (TREE_OPERAND (expr, 0)) == LABEL_DECL
-	      || !DECL_WEAK (TREE_OPERAND (expr, 0))))
-	return boolean_true_node;
+      if (decl_with_nonnull_addr_p (TREE_OPERAND (expr, 0)))
+	{
+	  warning (OPT_Waddress,
+		   "the address of %qD will always evaluate as %<true%>",
```

Comment:
Is this a useful message? It sounds confusing to me. Would 'the address of %qD is never null' or 'comparing the address of %qD to null will always evaluate as %<true%>' be better?

Reply:
It's the same message as `-Waddress` from gcc, so was just following by example here.

At this location, we'd be turned an address into a bool, i.e: `if (&foo) { ... }`.  There is no comparison to `null` here, that's what the other warning is for.

Reply:
> that's what the other warning is for.

In d-codegen, which takes care of `if (&foo is null)` and `if (&foo !is null)`.

Reply:
OK

Reply:
OK

Reply:
ok